docs/uri-parser-rewriter-proposal.md

```diff
@@ -0,0 +1,272 @@
+from the input string `"category eq 'electronics' or price gt 100"`.
+
+The benchmarks compare parsing the following filter expression: `"category eq 'electronics' or price gt 100"`.
+
+```md
+BenchmarkDotNet v0.14.0, Windows 11 (10.0.26100.2314)
+Intel Xeon W-2123 CPU 3.60GHz, 1 CPU, 8 logical and 4 physical cores
+.NET SDK 9.0.100
+  [Host]     : .NET 8.0.11 (8.0.1124.51707), X64 RyuJIT AVX-512F+CD+BW+DQ+VL
+  DefaultJob : .NET 8.0.11 (8.0.1124.51707), X64 RyuJIT AVX-512F+CD+BW+DQ+VL
+```
+
+| Method                                   | Mean         | Error      | StdDev     | Gen0   | Allocated |
+|----------------------------------------- |-------------:|-----------:|-----------:|-------:|----------:|
+| **ParseExpression_SlimExpressionLexer**      |     90.21 ns |   1.688 ns |   1.579 ns |      - |         - |
+
+See a sample implementation here: https://github.com/habbes/experiments/blob/master/ODataSlimUrlParserConcept/Lib/ExpressionLexer.cs.
+
+See a sample usage here: https://github.com/habbes/experiments/blob/master/ODataSlimUrlParserConcept/SlimParserTests/ExpressionLexerTests.cs
+
+#### Expression parsers and AST generation
+
+The current `UriQueryExpressionParser` recursively parses a sequence of token into a tree based on the `QueryToken` class.
+
+- Different subclasses of `QueryToken` represent different types of node.
+- We allocate substrings of the input to store text values in the `QueryToken`s.
+- Binary operator nodes have a reference to the left and right children and `QueryToken`.
+- The `LiteralToken` is used to represent, among others, parenthesized expressions (e.g. collection expressions). The expression is still in unparsed string form.
+
+This is where we propose a drastic change. Instead of nodes with direct reference to each other, we store nodes in a flat array.
+
+A node would have roughly the following structure:
+
+```csharp
+internal struct ExpressionNode
+{
+    public ExpressionNodeKind Kind { get; internal set; }
+    public ValueRange Range { get; internal set; }
+    internal int Count { get; set; }
+    internal int LastChild { get; set; }
+    public bool IsTerminal => Count == -1;
+}
+```
+
+And the expression parser will look roughly as follows:
+
+```csharp
+public ExpressionParser
+{
+    internal readonly ReadOnlyMemory<char> _source;
+    internal NodeList<ExpressionNode> _nodes;
+    
+    public static QueryNode Parse(ReadOnlyMemory<char> source)
+    {
+        var lexer = new ExpressionLexer(_source.Span);
+        int root = ParseExpression(ref lexer, ref _nodes, 0);
+        return QueryNode(this, root);
+    }
+}
+```
+
+The `ExpressioNode` is an internal representation of any kind of node in the "tree". This struct is not exposed publicly.
+Instead, we return a `QueryNode` struct which has the following structure:
+
+```csharp
+public struct QueryNode

```

Comment:
I don't see what prevents this from being a `ref struct` instead

Reply:
And I think we don't actually need the setters and we can make this `readonly` right?